### PR TITLE
sd-device: fix device_clone_with_db()

### LINKED ISSUE #27981: The old symbolic links are not deleted when the NVMe partition is renamed
### systemd version the issue has been seen with

systemd 253 (253-2728-g159f1b7)

### Used distribution

FusionOS-23_23.0.2.B002_x86_64

### Linux kernel version used

5.10.0-136.30.0.106.u42.fos23.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-udevd

### Expected behaviour you didn't see

ls -l  /dev/disk/by-partlabel/ |grep xxxx
lrwxrwxrwx 1 root root 15 Jun  9 08:19 xxxxx200 -> ../../nvme0n1p3

### Unexpected behaviour you saw

lrwxrwxrwx 1 root root 15 Jun  9 08:29  xxxxx106 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:29  xxxxx120 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx125 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx127 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:26  xxxxx13 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx131 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:26  xxxxx14 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:30  xxxxx146 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:26  xxxxx15 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx150 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx153 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx156 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx158 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx159 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx160 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx169 -> ../../nvme0n1p3
lrwxrwxrwx 1 root root 15 Jun  9 08:31  xxxxx173 -> ../../nvme0n1p3

### Steps to reproduce the problem

prepare: (nvme disk: create 40 partitions)
parted /dev/nvme0n1 --script mkpart primary1 ext4 103G 105G
parted /dev/nvme0n1 --script mkpart primary2 ext4 106G 108G
parted /dev/nvme0n1 --script mkpart primary3 ext4 109G 111G
parted /dev/nvme0n1 --script mkpart primary4 ext4 112G 114G
parted /dev/nvme0n1 --script mkpart primary5 ext4 115G 117G
...
parted /dev/nvme0n1 --script mkpart primary32 ext4 196G 198G
parted /dev/nvme0n1 --script mkpart primary33 ext4 199G 201G
parted /dev/nvme0n1 --script mkpart primary34 ext4 202G 204G
parted /dev/nvme0n1 --script mkpart primary35 ext4 205G 207G
parted /dev/nvme0n1 --script mkpart primary36 ext4 208G 210G
parted /dev/nvme0n1 --script mkpart primary37 ext4 211G 213G
parted /dev/nvme0n1 --script mkpart primary38 ext4 214G 216G
parted /dev/nvme0n1 --script mkpart primary39 ext4 217G 219G
parted /dev/nvme0n1 --script mkpart primary40 ext4 220G 222G

test1.sh: (change label)
for i in `seq 200`; do
parted /dev/nvme0n1 --script name 3 xxxxx$i;
sleep 2
done

test2.sh: (repeat add, remove)
for i in `seq 200`; do
parted /dev/nvme0n1 --script mkpart primary32 ext4 196G 198G
sleep 2
parted /dev/nvme0n1 --script print  |grep -w 198GB | awk '{print $1}' | xargs parted /dev/nvme0n1 --script rm
done

test3.sh:  (repeat add, remove)
for i in `seq 200`; do
parted /dev/nvme0n1 --script mkpart primary202 ext4 202G 204G
sleep 2
parted /dev/nvme0n1 --script print  |grep -w 204GB | awk '{print $1}' | xargs parted /dev/nvme0n1 --script rm
done

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
KERNEL[49574.218161] change   /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=change
MAJOR=259
MINOR=45             
in processing change event:
device_clone_with_db() {
	device_shallow_clone() -> device_set_devnum(dest, major=259, minor=45)
	device_read_uevent_file() --> change major:minor
	device_read_db_internal() --> read WRONG udev database file:run/udev/data/b259:15 (device_get_device_id RETURN b259:15) // introduce problem points
}

KERNEL[49574.318800] remove   /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=remove
MAJOR=259
MINOR=45

KERNEL[49574.320786] add      /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=add
MAJOR=259
MINOR=15


KERNEL[49576.297689] change   /devices/pci0000:e2/0000:e2:02.0/0000:e3:00.0/nvme/nvme0/nvme0n1/nvme0n1p14 (block)
ACTION=change
MAJOR=259
MINOR=15
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #27981.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/libsystemd/sd-device/device-private.c [hunk 1]
```diff
@@ -620,79 +620,63 @@ int device_get_devlink_priority(sd_device *device, int *ret) {
...
+int device_clone_with_db(sd_device *device, sd_device **ret) {
         _cleanup_(sd_device_unrefp) sd_device *dest = NULL;
-        const char *val = NULL;
+        const char *key, *val;
         int r;
 
         assert(device);
         assert(ret);
 
+        /* The device may be already removed. Let's copy minimal set of information that obtained through
```
Comment (bluca) [inline src/libsystemd/sd-device/device-private.c]: "that was obtained"

File: src/libsystemd/sd-device/device-private.c [hunk 2]
```diff
@@ -620,79 +620,63 @@ int device_get_devlink_priority(sd_device *device, int *ret) {
...
+         * netlink socket. */
+
         r = device_new_aux(&dest);
         if (r < 0)
                 return r;
 
+        /* Seal device to prevent reading the uevent file, as the device may be already removed. */
```
Comment (bluca) [inline src/libsystemd/sd-device/device-private.c]: "may have been already removed"

**Comment by bluca** [comment]
This is not changing API, so I think it doesn't need to block RC1? In can go in rc2 too

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
Thank you for the review. The suggestions are applied. Upgrading the green label.

> This is not changing API, so I think it doesn't need to block RC1? In can go in rc2 too

Maybe. But udevd is one of the critical component of systemd, hence I'd like to not change too much after -rc1. Feel free to drop the 'rc-blocker' tag if you cannot wait for CIs for this PR being finished.

**Pull request merged**